### PR TITLE
doc(pubsub): a custom thread pool example

### LINKED ISSUE #4586: Implement an example of setting your own background threads
And link this example from the `PublisherClient` and `SubscriberClient` documentation.

### PR DESCRIPTION
**Pull request by coryan**
Show how to setup a custom thread pool for `pubsub::Publisher` and
`pubsub::Subscriber`.

Fixes #4586

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/4726)

### CODE REVIEW

File: google/cloud/pubsub/samples/samples.cc [hunk 1]
```diff
@@ -234,6 +337,7 @@ void AutoRun(std::vector<std::string> const& argv) {
   auto generator = google::cloud::internal::MakeDefaultPRNG();
   auto topic_id = RandomTopicId(generator);
   auto subscription_id = RandomSubscriptionId(generator);
+  auto s2 = RandomSubscriptionId(generator);
```
Comment (devbww) [inline google/cloud/pubsub/samples/samples.cc:340]: Unused?

File: google/cloud/pubsub/samples/samples.cc [hunk 2]
```diff
@@ -221,6 +221,109 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
   (std::move(subscriber), std::move(subscription));
 }
 
+void CustomThreadPoolPublisher(std::vector<std::string> const& argv) {
+  namespace examples = ::google::cloud::testing_util;
+  if (argv.size() != 2 || (argv.size() == 1 && argv[0] == "--help")) {
```
Comment (devbww) [inline google/cloud/pubsub/samples/samples.cc:226]: Given that `lhs` is false when evaluating `rhs` in `lhs || rhs`, and that `!lhs` implies `!rhs` in this case, this guard is equivalent to:
```
if (argv.size() != 2) {
```
You might say the existing incantation is a more general pattern that allows for `s/2/1/` should the command only take one argument, but, if such an edit was made, I would probably then suggest changing
```
if (argv.size() != 1 || (argv.size() == 1 && argv[0] == "--help")) {
```
to just
```
if (argv.size() != 1 || argv[0] == "--help") {
```
In either case, simplicity should win, methinks.

File: google/cloud/pubsub/samples/samples.cc [hunk 3]
```diff
@@ -221,6 +221,109 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
...
+    cq.Shutdown();
+    for (auto& t : tasks) t.join();
+  }
+  //! [custom-thread-pool-publisher]
+  (argv.at(0), argv.at(1));
+}
+
+void CustomThreadPoolSubscriber(std::vector<std::string> const& argv) {
+  namespace examples = ::google::cloud::testing_util;
+  if (argv.size() != 2 || (argv.size() == 1 && argv[0] == "--help")) {
```
Comment (devbww) [inline google/cloud/pubsub/samples/samples.cc:274]: Ditto 1.

File: google/cloud/pubsub/samples/samples.cc [hunk 4]
```diff
@@ -221,6 +221,109 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
...
+  }
+  //! [custom-thread-pool-publisher]
+  namespace pubsub = google::cloud::pubsub;
+  using google::cloud::future;
+  using google::cloud::StatusOr;
+  [](std::string project_id, std::string topic_id) {
+    // Create our own completion queue to run the background activity, such as
+    // flushing the publisher.
+    google::cloud::CompletionQueue cq;
+    // Setup 1 or more of threads to service this completion queue, these must
```
Comment (devbww) [inline google/cloud/pubsub/samples/samples.cc:238]: s/, these/. These/

File: google/cloud/pubsub/samples/samples.cc [hunk 5]
```diff
@@ -221,6 +221,109 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
...
+    // remain running until all the work is done.
+    std::vector<std::thread> tasks;
+    std::generate_n(std::back_inserter(tasks), 4, [&cq] {
+      return std::thread(&google::cloud::CompletionQueue::Run, &cq);
```
Comment (devbww) [inline google/cloud/pubsub/samples/samples.cc:242]: We may not want to advocate for users taking addresses of our functions.  That is, use a lambda instead?

File: google/cloud/pubsub/samples/samples.cc [hunk 6]
```diff
@@ -221,6 +221,109 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
...
+  }
+  //! [custom-thread-pool-subscriber]
+  namespace pubsub = google::cloud::pubsub;
+  using google::cloud::future;
+  using google::cloud::StatusOr;
+  [](std::string project_id, std::string subscription_id) {
+    // Create our own completion queue to run the background activity, such as
+    // flushing the publisher.
+    google::cloud::CompletionQueue cq;
+    // Setup 1 or more of threads to service this completion queue, these must
```
Comment (devbww) [inline google/cloud/pubsub/samples/samples.cc:286]: Ditto 2.

File: google/cloud/pubsub/samples/samples.cc [hunk 7]
```diff
@@ -221,6 +221,109 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
...
+    // remain running until all the work is done.
+    std::vector<std::thread> tasks;
+    std::generate_n(std::back_inserter(tasks), 4, [&cq] {
+      return std::thread(&google::cloud::CompletionQueue::Run, &cq);
```
Comment (devbww) [inline google/cloud/pubsub/samples/samples.cc:290]: Ditto 3.

File: google/cloud/pubsub/samples/samples.cc [hunk 8]
```diff
@@ -221,6 +221,109 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
...
+            if (++count == 4) received_message.set_value();
+          }
+          std::move(h).ack();
+        });
+    // Cancel the subscription as soon as the first message is received.
+    received_message.get_future()
+        .then([&result](google::cloud::future<void>) { result.cancel(); })
+        .get();
+    // Report any final status, blocking until the subscription loop completes,
+    // either with a failure or because it is canceled.
```
Comment (devbww) [inline google/cloud/pubsub/samples/samples.cc:315]: Opt: s/is canceled/was canceled/ perhaps?

File: google/cloud/pubsub/samples/samples.cc [hunk 9]
```diff
@@ -221,6 +221,109 @@ void Subscribe(google::cloud::pubsub::Subscriber subscriber,
...
+    std::cout << "Message count = " << count << ", status=" << result.get()
```
Comment (devbww) [inline google/cloud/pubsub/samples/samples.cc:316]: Nitty nit: `"="` or `" = "`.

Comment (coryan) [reply google/cloud/pubsub/samples/samples.cc]: Fixed.

File: google/cloud/pubsub/samples/samples.cc (hunk 8 above)
Comment (coryan) [reply google/cloud/pubsub/samples/samples.cc]: Done.

File: google/cloud/pubsub/samples/samples.cc (hunk 7 above)
Comment (coryan) [reply google/cloud/pubsub/samples/samples.cc]: Fixed.

File: google/cloud/pubsub/samples/samples.cc (hunk 6 above)
Comment (coryan) [reply google/cloud/pubsub/samples/samples.cc]: Also done.

File: google/cloud/pubsub/samples/samples.cc (hunk 5 above)
Comment (coryan) [reply google/cloud/pubsub/samples/samples.cc]: Hmmm... Okay, I understand why.  Done.

File: google/cloud/pubsub/samples/samples.cc (hunk 4 above)
Comment (coryan) [reply google/cloud/pubsub/samples/samples.cc]: Done, thanks.

File: google/cloud/pubsub/samples/samples.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/pubsub/samples/samples.cc]: Doh! Fixed.

File: google/cloud/pubsub/samples/samples.cc (hunk 1 above)
Comment (coryan) [reply google/cloud/pubsub/samples/samples.cc]: Yes, removed.

**Comment by coryan** [comment]
Rebased and fixed clang-tidy problems.

**Pull request merged**